code/modules/mob/dead/observer/observer.dm

```diff
@@ -283,8 +283,14 @@ Works together with spawning an observer, noted above.
 /mob/proc/ghostize(can_reenter_corpse = TRUE)
 	if(key)
 		if(key[1] != "@") // Skip aghosts.
-			if(HAS_TRAIT(src, TRAIT_CORPSELOCKED) && can_reenter_corpse) //If you can re-enter the corpse you can't leave when corpselocked
-				return
+			if(HAS_TRAIT(src, TRAIT_CORPSELOCKED))
+				if(can_reenter_corpse) //If you can re-enter the corpse you can't leave when corpselocked
+					return
+				else //otherwise, continues the ghosting process
+					if(isethereal(usr)) //and cancels crystallization process for ethereals
+						var/mob/living/carbon/human/ethereal = usr
+						var/obj/item/organ/heart/ethereal/ethereal_heart = ethereal.getorganslot(ORGAN_SLOT_HEART)
+						ethereal_heart.stop_crystalization_process(ethereal)
```

Comment:
non-ethereals can be revived by ethereal hearts, IIRC, and ethereals can get their hearts replaced with mechanical or human ones. the isethereal() check here should instead be replaced by a check to see if the humanoid's heart is an ethereal heart.